_published/0026-v2-go-modules-release.md

```diff
@@ -0,0 +1,180 @@
+title: "RFC-0026: v2+ Go modules release" # Update this with PR number/ID and feature name. Use leading zeros
+rfc_author_username: jilleJr
+rfc_author_name: Kalle Fagerberg # Or same as username, if you wish
+
+# Leave these. Collaborator changes this before merging
+impl_issue_id: 0
+impl_issue_repo: iver-wharf/wharf-api
+last_modified_date: YYYY-MM-DD
+---
+
+# {{page.title}}
+
+- RFC PR: [iver-wharf/rfcs#{{page.pr_id}}](https://github.com/iver-wharf/rfcs/pull/{{page.pr_id}})
+- Feature name: `{{page.rfc_feature_name}}`
+- Author: {{page.rfc_author_name}} ([@{{page.rfc_author_username}}](https://github.com/{{page.rfc_author_username}}))
+- Implementation issue: [{{page.impl_issue_repo}}#{{page.impl_issue_id}}](https://github.com/{{page.impl_issue_repo}}/issues/{{page.impl_issue_id}})
+- Implementation status: ![GitHub issue state](https://img.shields.io/github/issues/detail/state/{{page.impl_issue_repo}}/{{page.impl_issue_id}}?{{page.shields_io_query_params}})
+
+## Summary
+
+Go modules, which is the package management strategy we use in Wharf, has
+strict rules on how to publish v2+ releases. We have not been Go-modules
+compliant in some of our repositories. This RFC suggest one of the many
+alternative solutions for being compliant, with its up- and downsides.
+This is an important decision, as it induces a lot of headaches if we do it
+wrong and have to switch to a different strategy later on.
+
+> There's an ironic trivia fact here: The documentation for Go modules is
+> longer than the full [Go language specification](https://go.dev/ref/spec),
+> but worse is that the Go modules resources are spread out across blog posts,
+> wiki's, and documentation sites.
+
+This is based on a conversation that popped up in a PR comment thread:
+<https://github.com/iver-wharf/wharf-api-client-go/pull/29#discussion_r766513730>
+
+## Motivation
+
+The wharf-api is an example of a version 2 or higher Go module we've release.
+Trying to import packages from for example v4.2.0 results in an error:
+
+```console
+$ go get github.com/iver-wharf/wharf-api/pkg/orderby@v4.2.0
+go get github.com/iver-wharf/wharf-api/pkg/orderby@v4.2.0:
+  github.com/iver-wharf/wharf-api@v4.2.0:
+    invalid version:
+      module contains a go.mod file, so major version must be compatible:
+        should be v0 or v1, not v4
+```
+
+*(I've wrapped the above console output just to be more readable)*
+
+## Explanation
+
+We release new major versions by updating the module path in `go.mod` and all
+related imports in `*.go` files.
+
+Example:
+
+```diff
+-module github.com/iver-wharf/wharf-api
++module github.com/iver-wharf/wharf-api/v5
+
+ go 1.16
+```
+
+```diff
+ package main
+
+ import (
+-    "github.com/iver-wharf/wharf-api/pkg/model/database"
++    "github.com/iver-wharf/wharf-api/v5/pkg/model/database"
+ )
+```
+
+### Counter-recommendation decisions
+
+Contrary to some of the official Go recommendations, we do not apply any of the
```

Comment:
This phrasing confuses me, so we do not apply **no** separate branches?

We don't use separate branches for major versions right? The double negation makes it sound like we do use separate branches. But maybe that's just me.

Reply:
No you're completely right. This paragraph is a mess with its double negation.

I've changed it:

From:

> Contrary to some of the official Go recommendations, we do not apply any of the following strategies:

To:

> Contrary to some of the official Go recommendations, our practices are: